zilliqa/src/pool.rs

```diff
@@ -289,14 +309,40 @@ impl TransactionPool {
             Self::add_to_gas_index(&mut self.gas_index, &txn);
         }
 
+        // If this is a transaction created at this node, add it to broadcast vector
+        if !from_broadcast {
+            self.store_broadcast_txn(txn.tx.clone());
+        }
+
         debug!("Txn added to mempool. Hash: {:?}, from: {:?}, nonce: {:?}, account nonce: {account_nonce}", txn.hash, txn.signer, txn.tx.nonce());
 
         // Finally we insert it into the tx store and the hash reverse-index
         self.hash_to_index.insert(txn.hash, txn.mempool_index());
         self.transactions.insert(txn.mempool_index(), txn);
+
         TxAddResult::AddedToMempool
     }
 
+    fn store_broadcast_txn(&mut self, txn: SignedTransaction) {
+        self.transactions_to_broadcast.push(txn);
+        if self.first_broadcast_txn_add_time == SystemTime::UNIX_EPOCH {
+            self.first_broadcast_txn_add_time = SystemTime::now();
+        }
+    }
+
+    pub fn pull_txns_to_broadcast(&mut self) -> Result<Vec<SignedTransaction>> {
+        const MAX_TIME_SINCE_FIRST_INSERT: Duration = Duration::milliseconds(50);
+        const MAX_BATCH_SIZE: usize = 20;
+
+        if self.transactions_to_broadcast.len() > MAX_BATCH_SIZE
```

Comment:
Instead of having a minimum batch size, and needing to check for 'staleness', I would probably have a maximum batch size limit (just in case there are too many) and pull whatever is available up to that max size, at the point to broadcast.